pkg/sti/build.go

```diff
@@ -73,40 +78,59 @@ func (b *Builder) Build() (*Result, error) {
 	if err != nil {
 		return nil, err
 	}
-	if bh.Request().incremental {
+	if bh.Request().Incremental {
 		glog.V(1).Infof("Existing image for tag %s detected for incremental build.", bh.Request().Tag)
 	} else {
 		glog.V(1).Infof("Clean build will be performed")
 	}
 
 	glog.V(2).Infof("Performing source build from %s", bh.Request().Source)
-	if bh.Request().incremental {
+	if bh.Request().Incremental {
 		if err = bh.saveArtifacts(); err != nil {
 			glog.Warning("Error saving previous build artifacts: %v", err)
 			glog.Warning("Clean build will be performed!")
 		}
 	}
 
-	if err = bh.execute("assemble"); err != nil {
+	glog.V(1).Infof("Building %s", bh.Request().Tag)
+	err = bh.execute(api.Assemble)
+	if e, ok := err.(errors.ContainerError); ok && bh.wasExpectedError(e.ExpectedError) {
+		glog.Warningf("Image %s does not have tar! Performing additional build to add the scripts and sources.",
+			bh.Request().BaseImage)
+		if err := bh.build(); err != nil {
+			return nil, err
+		}
+		glog.V(2).Infof("Building %s using sti-enabled image", bh.Request().Tag)
+		if err := bh.execute(api.Assemble); err != nil {
+			return nil, err
+		}
+	} else if err != nil {
 		return nil, err
 	}
 
 	return bh.Result(), nil
 }
 
-func (h *buildHandler) PostExecute(containerID string, cmd []string) error {
+func (h *buildHandler) wasExpectedError(text string) bool {
```

Comment:
Unfortunately current logic hides the from the build part the docker internals,  my initial idea here was to log the entire stderr and return that to user with `ContainerError`. But @csrwng complained about that idea, so I came up with this solution, but frankly said I'm partially satisfied with this. So any suggestions to make it better are welcome. In the mean time I've added necessary comment to that method.